notify/smtp.sh

```diff
@@ -2,14 +2,330 @@
+    if [ -n "$smtp_output" ]; then
+      _err "$smtp_output"
+    fi
+    return 1
+  fi
+
+  # Save config only if send was successful:
+  _saveaccountconf_mutable SMTP_BIN "$SMTP_BIN"
+  _saveaccountconf_mutable SMTP_FROM "$SMTP_FROM"
+  _saveaccountconf_mutable SMTP_TO "$SMTP_TO"
+  _saveaccountconf_mutable SMTP_HOST "$SMTP_HOST"
+  _saveaccountconf_mutable SMTP_PORT "$SMTP_PORT"
+  _saveaccountconf_mutable SMTP_SECURE "$SMTP_SECURE"
+  _saveaccountconf_mutable SMTP_USERNAME "$SMTP_USERNAME"
+  _saveaccountconf_mutable SMTP_PASSWORD "$SMTP_PASSWORD"
+  _saveaccountconf_mutable SMTP_TIMEOUT "$SMTP_TIMEOUT"
+
+  return 0
+}
+
+# Send the message via curl using _SMTP_* variables
+_smtp_send_curl() {
+  # curl passes --mail-from and --mail-rcpt directly to the SMTP protocol without
+  # additional parsing, and SMTP requires addr-spec only (no display names).
+  # In the future, maybe try to parse the addr-spec out for curl args (non-trivial).
+  if _email_has_display_name "$_SMTP_FROM"; then
+    _err "curl smtp only allows a simple email address in SMTP_FROM."
+    _err "Change your SMTP_FROM='$SMTP_FROM' to remove the display name."
+    return 1
+  fi
+  if _email_has_display_name "$_SMTP_TO"; then
+    _err "curl smtp only allows simple email addresses in SMTP_TO."
+    _err "Change your SMTP_TO='$SMTP_TO' to remove the display name(s)."
+    return 1
+  fi
+
+  # Build curl args in $@
+
+  case "$_SMTP_SECURE" in
+  none)
+    set -- --url "smtp://${_SMTP_HOST}:${_SMTP_PORT}"
+    ;;
+  ssl)
+    set -- --url "smtps://${_SMTP_HOST}:${_SMTP_PORT}"
+    ;;
+  tls)
+    set -- --url "smtp://${_SMTP_HOST}:${_SMTP_PORT}" --ssl-reqd
+    ;;
+  *)
+    # This will only occur if someone adds a new SMTP_SECURE option above
+    # without updating this code for it.
+    _err "Unhandled _SMTP_SECURE='$_SMTP_SECURE' in _smtp_send_curl"
+    _err "Please re-run with --debug and report a bug."
+    return 1
+    ;;
+  esac
+
+  set -- "$@" \
+    --upload-file - \
+    --mail-from "$_SMTP_FROM" \
+    --max-time "$_SMTP_TIMEOUT"
+
+  # Burst comma-separated $_SMTP_TO into individual --mail-rcpt args.
+  _to="${_SMTP_TO},"
+  while [ -n "$_to" ]; do
+    _rcpt="${_to%%,*}"
+    _to="${_to#*,}"
+    set -- "$@" --mail-rcpt "$_rcpt"
+  done
+
+  _smtp_login="${_SMTP_USERNAME}:${_SMTP_PASSWORD}"
+  if [ "$_smtp_login" != ":" ]; then
+    set -- "$@" --user "$_smtp_login"
+  fi
+
+  if [ "$_SMTP_SHOW_TRANSCRIPT" = "True" ]; then
+    set -- "$@" --verbose
+  else
+    set -- "$@" --silent --show-error
+  fi
+
+  raw_message="$(_smtp_raw_message)"
+
+  _debug2 "curl command:" "$_SMTP_BIN" "$*"
+  _debug2 "raw_message:\n$raw_message"
+
+  echo "$raw_message" | "$_SMTP_BIN" "$@"
+}
+
+# Output an RFC-822 / RFC-5322 email message using _SMTP_* variables
+_smtp_raw_message() {
+  echo "From: $_SMTP_FROM"
+  echo "To: $_SMTP_TO"
+  echo "Subject: $(_mime_encoded_word "$_SMTP_SUBJECT")"
+  if _exists date; then
+    echo "Date: $(date +'%a, %-d %b %Y %H:%M:%S %z')"
```

Comment:
please don't use `date` command.

please use the `_utc_date()` function and `sed/tr` instead.

Reply:
The email Date header must be in the format specified by [RFC 2822 Section 3.3](https://tools.ietf.org/html/rfc2822#section-3.3):
> Date: Mon, 15 Feb 2021 00:03:12 -0800

1. The month name ("Feb") must be a three-character English abbreviation; numeric months are not permitted
2. The weekday ("Mon") is optional, but if present must be a three-character English abbreviation; as a practical matter weekday almost always seems to be included, and omitting it could break some (poorly implemented) email clients or scripts
3. The time zone "SHOULD express local time" (though UTC is permitted)

Some options:
A. Ignore recommendations 2 and 3, and try to hack something together with `sed` on `_utc_date()` to generate month abbreviations for requirement 1. (This seems really complicated compared to just including `%b` in a `date` format specifier.)
B. Add a new `_rfc2822_date()` function to abstract this, let it call `date` like here. (Also, I notice it should override `LC_TIME=C` to force required English abbreviations.)
C. Don't add a Date header to the generated message, and hope that the SMTP server adds one for us. (Many will.)

Please let me know which you'd prefer. I can implement option B or C; if option A is preferred, perhaps you could suggest the sed expression.

Reply:
I'd like option B more.

please make sure `_rfc2822_date` can return exact the same format on Linux, Mac, BSD, and Solaris.


If you need BSD vm and Solaris VM, see:

https://github.com/vmactions/shell-solaris

https://github.com/vmactions/shell-freebsd


you can also implement `_rfc2822_date ` based on the output of `_utc_date`,  just  `sed` and `tr`it.

Reply:
> implement `_rfc2822_date` based on the output of `_utc_date`, just `sed` and `tr` it

Sorry, this is a little beyond my skills. How would I use `sed` and `tr` to transform `2021-02-16 17:14:43` into `Tue, 16 Feb 2021 17:14:43 +0000`? (I know how to rearrange pieces of text, but not how to change `02` into `Feb` and `2021-02-16` into `Tue`.)

Reply:
(I've added `_rfc2822_date` and tested on the various platforms; would still need help with `sed` if you want that changed.)